runtime/gc_vlhgc/WriteOnceCompactor.cpp

```diff
@@ -743,6 +743,14 @@ MM_WriteOnceCompactor::evacuatePage(MM_EnvironmentVLHGC *env, void *page, J9MM_F
 				}
 				if(_extensions->objectModel.isIndexable(newLocation)) {
 					updateInternalLeafPointersAfterCopy((J9IndexableObject*) newLocation, (J9IndexableObject*) objectPtr);
+#if defined(J9VM_ENV_DATA64)
```

Comment:
Let's have this opportunity to make code a bit more readable and more consistent (with OMR compactor), and try introduce WOC::preObjectMove/postObjectMove  methods (move everything about remembering, hashing etc. in them)

You may want to rely on objectMode.preMove/postMove, but be completely sure they do the same what we do here. They will be marginally slower (since accessing thread struct, rather than stack), but this is anyhow relatively infrequent/slow path, so I'm fine with it.

If not possible then, you'll have to extend signatures of WOC::preObjectMove/postObjectMove to pass out/in hash/move flags.

And now to be really annoying, I'll say that I don't like that movedObjectHashCode struct is in OMR Thread , while all other hash/move knowledge is very OpenJ9 specific. The struct should be in OpenJ9's Thread (what would require that objectMode.preMove/postMove deal with Environement and OMR thread, and they extract OpenJ9's Thread from Env). What do you think? Even if we want to change that, perhaps it would be cleaner to do separately!?




unlike in OMR (which allows different language to have different actions/flags for hashing & movement), you don't have to rely on objectModel.preMove/postMove

Reply:
I see what you mean, from a quick look it seems that it would be possible to substitute those calls with objectModel.preMove/postMove . I haven't found any usages of `movedObjectHashCodeCache` in OMR side either so moving it into J9VMThread does make sense. If we introduce `struct movedObjectHashCode movedObjectHashCodeCache;` in J9VMThread in this PR, we could just remove the equivalent one in OMR with the OMR PR.

Reply:
I have good and bad news. objectModel.preMove/postMove do exactly as what we want here hence we can substitute the code before and after `memmove` with those. I've also created `preObjectMove` and `postObjectMove` to be consistent with OMR compactor. The bad news is that we can substitute `OMR_VMThread` with `J9VMThread` and transfer `movedObjectHashCode` over; however, that would require that we update the objectModel.preMove/postMove API to accept `J9VMThread` instead of `OMR_VMThread`. With that, OMR compactor objectModel.preMove/postMove API complains that it must also use `J9VMThread`, which would lead us to also updating OMR compactor to use `J9VMThread` which is something we don't want? 

I tried to get access to `J9VMThread` through `OMR_VMThread` but I don't think that's possible.

Reply:
Right, but perhaps this is still doable..... 

movedObjectHashCode is used only within GC code, so it's even better to be inside GC (and J9) specific struct MM_EnvironmentDelegate (or GC_Environment? - honestly, don't know why they are separate). And also declared somewhere in GC specific file. 

objectModel.preMove/postMove would then stay with OMR specific parameter (OMR_VMThread or EnvironmentBase), from where we know how to access language specific delegates.

Reply:
I see, I think we can get access to those via `OMR_VMThread`, looking into it

Reply:
We can actually access `J9VMThread` from `OMR_VMThread` via member `_language_vmthread` (should have looked harder before). So I'll put it under `J9VMThread` and since we cannot name it the same and this is J9/GC specific I was thinking about `J9movedObjectHashCode` or maybe `GCmovedObjectHashCode`?

Reply:
I still prefer it in EnvDelegate. It's only used in GC code.

Reply:
then even change objectModel.preMove/postMove to pass EnvBase (again same argument, we don't need anything from OMRVMThread, everything is in GC structs).

For that, to break mutual dependancy with OMR you'll need to clone objectModel.preMove/postMove, and remove the old ones with a followup J9 PR.

Reply:
Now in order to get access to EnvDelegate we need to go through EnvBase, but EnvDelegate member in EnvBase is protected and defined in OMR side (it seems this is the only way we can access EnvDelegate). In that case we would need to update OMR side first before we can access `movedObjectHashCode` inside EnvDelegate, unless I'm missing something?

Reply:
you can put the struct in GC_Environment component of EnvDelegate, that seems to have a public getter in EnvBase:

	MMINLINE GC_Environment *getGCEnvironment() { return _delegate.getGCEnvironment(); }

Reply:
Everything is good with this approach except that I started having issues when trying to inline the new objectModel.preMove/postMove . Now that EnvBase is one of the parameters, objectModel.preMove/postMove bodies must go in the source cpp file instead of the header file. I need to do that because I can't include EnvBase from the header file; I get several `invalid use of incomplete type` error which are probably due to cyclic imports. I've had this issue in the past (trying to include `EnvironmentBase.hpp` in `ObjectModel.hpp`). And the inline error I get for not having the body in header file is:

```
error: inlining failed in call to always_inline 'void GC_ObjectModel::preMove...': function body not available
```
What would you recommend going forward? Should I deal with these cyclic reference issue?

Reply:
We are finally realizing why movedObjectHashCode struct (that is language specific) put into an OMR thread local struct, to start with. Without trying to understand everything, there are cycle references between EnvBase (what includes ExtBase) and ObjectModel (that is embedded in ExtBase). OMR thread local struct is out of that loop and also can be obtained by caller before we enter the loop by calling objectModel.preMove/postMove....

We might give up to this part of changes, afterall. But there might be some hope if we take preMove/postMove from ObjecdtModel into EnvBase  (even if that works, it's generally against the trend where we try to leave EnvBase lean)

Reply:
I think I got something going, followed by your advice. I’ve included preMove/postMove inlined implementation in `EnvironmentDelegate`, and have the according wrappers at `EnvironmentBase`. Now `struct movedObjectHashCode` resides in `GC_Environment`. Everything works but in order to this correctly (without having mutual dependency with OMR) I think we would need a 4 way set of PRs since we need the wrappers in place in `EnvironmentBase` which is in OMR side:
1. This PR first, where we introduce preMove/postMove inlined in `EnvironmentDelegate`
2. OMR PR https://github.com/eclipse/omr/pull/5795 where we introduce wrappers at `EnvironmentBase`. At this point we can update OMR compactor to use the new preMove/postMove from `EnvironmentDelegate`.
3. Another OpenJ9 PR that will update `WriteOnceCompactor` preMove/postMove to use the new `EnvironmentBase` preMove/postMove. And at this point we can remove ObjectModel.preMove/postMove. 
4. Last OMR PR where we remove the obsolete `struct movedObjectHashCode`. We can probably make this a beginner’s issue?

How does this sound?

Reply:
Yes, this is this looks good. 

But then, while we definitely did cleaner separation between OMR/OpenJ9, we did violate a tacit rule that Env/Ext should be lean. I'll leave it up to you, what you prefer more (or find a better solution).

If we stick with this, I think it would be ok to cache ExtBase in EnvDelegate (it already caches EnvBase, from which you can get ExtBase, but caching it explicitly is even better).

Reply:
Sounds good, let me give another thought to this and see if this is indeed better than having pre/postMove in ObjectModel

Reply:
I found another possible place to put preMove/postMove which is CompactSchemeFixupObject.hpp but it would be out of scope and no straight forward way to access unless we make them static methods. Given our circumstances/constraints and keeping the separation between OMR/OpenJ9, placing preMove/postMove in `EnvironmentDelegate` seems to be the best approach. It feels like preMove/postMove will be a little bit out of place (inside `EnvironmentDelegate`) but I don't see any other reasonable location to place it. The other 2 options would be:  
- leave things as is, which is to keep preMove/postMove inside ObjectModel  we'll need to keep `OMR_VMThread` and `movedObjectHashCode` in OMR side which is the main reason we're doing this. 
- Move `movedObjectHashCode` to J9VMThread and keep preMove/postMove inside ObjectModel, which is slightly better than current implementation

Reply:
The current state of the PR reflects approach **[1]** from the above comment.

Reply:
We do have a pair of similar methods in EnvBase already: save/restoreObject, that deal with whole objects before/after GC in some special cases. We do similar things but with specific object flags before after object movement. So it's not totally out of place, although not perfect fit either.

I'm not totally convinced moving getting CompactSchemeFixupObject involved would help, although I have not put much thought. 

Anyhow, I'm ok with the current approach, but if you prefer [2] or [3] more, I'd probably be with them too.

Reply:
CompactSchemeFixupObject wouldn't help at all, I was brainstorming. And sounds good, that's true I did not consider save/restoreObject before. Okay then let's stick to this approach, I'll address your latest comments where I'll update preMove/postMove to the more appropriate names and add some comments there as well